### PR TITLE
Implement thread pool and affinity classes

### LINKED ISSUE #19: [RFC] Introduction of a thread affinity helper type
C++ does not currently implement any standard way to interact with thread affinity, making writing programs that need thread pinning an exercise in using the raw OS-specific threading APIs.

[pcat](https://github.com/dragonmux/pcat) addressed this by implementing a platform-specific thread affinity type (`affinity_t`) for [Linux](https://github.com/dragonmux/pcat/blob/main/src/linux/affinity.hxx) and [Windows](https://github.com/dragonmux/pcat/blob/main/src/windows/affinity.hxx).

It seems this would be a good inclusion into substrate

### LINKED ISSUE #20: [RFC] Introduction of a thread pool
👋 

One of the key selling points of Qt vs. STL at $work is the availability of a cross-platform thread pool. I think @dragonmux has an implementation in one of her projects.

A key feature would be to accept fully type-erased lambdas à la `QRunnable` (which fully hides, unlike `std::function`, the underlying function's signature).

### PR DESCRIPTION
**Pull request by amyspark**
Hi @dragonmux and @lethalbit,

As promised, I ported the thread affinity and pool classes from 🐉 's pcat utility. I also added MSVC, MinGW and macOS/iOS support. You'll also find some assorted extras to ensure we cut down on the number of warnings and incompatibilities, in particular in the Windows cases.

Let me know what you think 🐱 

Fixes #19  
Fixes #20

### CODE REVIEW

**Review by dragonmux — CHANGES_REQUESTED** [review]
Looking good, the only review points we spotted are in the affinity implementation itself and are easy fix items.

Think we should still try and provide a second kind of thread pool in the future that's less specialised to what pcat needed, by making use of std::promise/std::future to allow any work to be queued and the answers later dequeued, but this can be the subject of another PR and a new issue ticket

File: impl/affinity.cxx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+			if ((!threadCount || processors.size() < threadCount) &&
+			(pinning.empty() || std::find(pinning.cbegin(), pinning.cend(), i) != pinning.cend()))
+				processors.emplace_back(i);
+			CPU_CLR(i, &affinity);
+		}
+		else if (!CPU_COUNT(&affinity))
+			break;
+	}
+#elif defined(_POSIX_THREADS) && defined(__APPLE__)
+	const auto core_count = []() -> unsigned int {
```
[the remark this answers is not in the document]
Comment (amyspark) [reply impl/affinity.cxx:79]: https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency says that `hardware_concurrency` returns an implementation-dependent `unsigned int`, not `uint32_t` unfortunately. The rest I can address in the next patchset.

**Comment by amyspark** [comment]
Also something to think about, `pinning` doesn't look to be addressed other than with `std::find` which is O(n) in all cases. Perhaps that could be turned into a `std::set` or `std::unordered_set`?

Unfortunately, we can't do the same to `processors` because the affinity mask is retrieved through indexing >.<

**Comment by dragonmux** [comment]
yeah, switching the pinning stuff to a std::set would be just fine with us - similarly, if you think the available processor set can be better represented with a std::bitset, then you'd be welcome to switch it over

**Review by dragonmux — COMMENTED** [review]
Happy with the naming fixes, leaving just a formatting issue in impl/affinity.cxx to address, and some build failures to look into (an errant auto on the line before the suggested change?)

**Review by dragonmux — APPROVED** [review]

**Pull request merged**